### DISCUSSION TITLE
Why using map combinator on delegate excludes it from list of tokens?

### DISCUSSION THREAD

**Discussion by gumanoid** [h0tk3y/better-parse#53] [Q&A] [answered]
This code successfully parses any text:
```
object Wut : Grammar<List<String>>() {
    val literal by regexToken(".{1,3}")
    override val rootParser by oneOrMore(literal use { text })
}
```
However, if I move `use { text }` to definition of literal:
```
object Wut : Grammar<List<String>>() {
    val literal by regexToken(".{1,3}") use { text }
    override val rootParser by oneOrMore(literal)
}
```
 it complains "the token list should not be empty".

Why position of `use { text }` matters? For me this seems to break referential transparrency.

I'm also puzzled why token list is "empty" in second case? I've explicitly told it to use `literal`

**Comment by h0tk3y** [comment, marked as answer]
Defining tokens as separate declarations inside the grammar is needed for collecting the tokens and building a tokenizer on top of them. If you use tokens "inline" in other parsers, the grammar's current implementation won't be able to [register the tokens](https://github.com/h0tk3y/better-parse/blob/1b88de85978c17bb8d197d631319a220dc188d69/src/commonMain/kotlin/com/github/h0tk3y/betterParse/grammar/Grammar.kt#L45-L51) and use them in a tokenizer.

So, the workaround is to always declare tokens as delegated properties inside a grammar. 

Given that the set of tokens and their order really matters in parsing, following the convention of defining tokens as delegated properties also makes it more clear as to what tokens are there and in which order.

The grammar implementation could potentially be improved in some ways that would make best effort to enforce registering tokens as properties.

If you want, you can customize the logic for registering tokens in your grammar and override the grammar's [`val tokens: List<Token>`](https://github.com/h0tk3y/better-parse/blob/1b88de85978c17bb8d197d631319a220dc188d69/src/commonMain/kotlin/com/github/h0tk3y/betterParse/grammar/Grammar.kt#L25) so that it returns the actual set of the tokens that it needs to use in tokenizing the input.

**Reply by gumanoid** [reply]
First of all, thanks for clarifications.
I've overlooked that `*token` functions are returning `Token`, and map combinator returns `Parser`

Token order really matters to parsing, indeed. But I doubt that it should be mapped to order of field declarations. Order of field/method declarations inside of a class/object, in general, can be arbitrary and should not affect code behavior. Thus it is unexpected to average programmer that field reordering could break something.
The concept of ordering maps naturally to `List<T>`. For me, the code that explicitly declares list of tokens in order, is much clearer. Maybe overriding `val tokens: List<Token>` should be mandatory just like `rootParser`?

**Reply by gumanoid** [reply]
One more thing that I can't get is: parser combinators is purely functional concept. Why mess with OOP world here by leveraging delegation at all?
